### DISCUSSION TITLE
time_periodic

### DISCUSSION THREAD

**Discussion by ignasivalles** [Parcels-code/Parcels#1475] [Q&A] [answered]
Hi, I'm running parcels looping over a period of time so I use the option time_periodic. I understand the philosophy, but I'm not really sure what it does, since by mistake, I added a period of time longer than the velocity field (I was confused by the time period I want for my run). The code didn't give any problem, as "time_periodic has to be equal to or longer than the time in the fieldset", so when it is longer, what's going on? When the program loops, do the particles jump from the last time step to the first anyway?

(Btw, if I'm not wrong, I've seen that the option time_periodic=True does not work anymore, but it still appears in the tutorial documentation: https://docs.oceanparcels.org/en/latest/examples/tutorial_timestamps.html)

thank you!

**Comment by erikvansebille** [comment, marked as answer]
Hi @ignasivalles, thanks for asking this question. I must say that we don't often (never?) use `time_periodic` in my group, so I'm not entirely sure whether it still works as originally intended. 

As for the expected length, I must admit that this is very confusing, but I think the point is that the last time can be duplicated, see the `self.grid._add_last_periodic_data_timestep` variable on line 203 in the code below:
https://github.com/OceanParcels/parcels/blob/7e47aa3c26fc25072921c6d9050cd0d58edc0c47/parcels/field.py#L197-L205

FYI, the code that then converts the periodic time is below:
https://github.com/OceanParcels/parcels/blob/7e47aa3c26fc25072921c6d9050cd0d58edc0c47/parcels/field.py#L1163-L1193

So in summary, I think that the expectation is that the value of `time_periodic` is exactly the same as the dataset, and that the only freedom you have is whether the last time is a duplicate of the first time or not. 
If you want to periodically loop over a shorter time_window, then you should instead use fewer hydrodynamic snapshots when you create your FieldSet.

Oh and good catch about the old statement in the tutorial, I'll fix that immediately

**Comment by ignasivalles** [comment]
Hi again. Regarding that question, I would like to point out that when you define `time_periodic` with the `timedelta` function (e.g., `time_periodic = datetime.timedelta(days=6393)`), surprisingly, the program crashes. However, it works fine when you simply add an integer number in seconds.

**Reply by erikvansebille** [reply]
Hmm, that is strange because this code should convert a `datetime.timedelta` to seconds
https://github.com/OceanParcels/parcels/blob/3cf8b2ac7fe0b7f4da8a6dc7536b4dff85d8c183/parcels/field.py#L198-L199

Could you specify what code caused the 'crash', and what exactly was the error?

**Reply by ignasivalles** [reply]
Thanks, @erikvansebille, for your reply. Yes, it's strange, it appears to be a memory issue. However, when I run it with just the number in seconds, it works fine. The terminal was killed with no errors displayed. Here is a snippet of my code. It looks like it gets stuck in `FieldSet.from_nemo` function. Please note that I'm working with a substantial amount of data (>2TB).

```
Ufiles  =  sorted(glob(data_file_path+"U2/GLORYS12V1-gridU-*.nc"))
Vfiles  =  sorted(glob(data_file_path+"V/GLORYS12V1-gridV-*.nc")) 
Wfiles  =  sorted(glob(data_file_path+"W2/GLORYS12V1-gridW-*.nc")) 
Tfiles = sorted(glob(data_file_path+"T2/GLORYS12V1-gridT-*.nc"))
Sfiles = sorted(glob(data_file_path+"S2/GLORYS12V1-gridS-*.nc"))

mesh_mask = data_file_path+"GLORYS12V1-mesh_mask.nc"

filenames = {'U': {'lon':mesh_mask, 'lat':mesh_mask, 'depth':Wfiles[0],'time':Ufiles, 'data':Ufiles}, 
             'V': {'lon':mesh_mask, 'lat':mesh_mask, 'depth':Wfiles[0],'time':Vfiles, 'data':Vfiles}, 
             'W': {'lon':mesh_mask, 'lat':mesh_mask, 'depth':Wfiles[0],'time':Wfiles, 'data':Wfiles},
             'T': {'lon':mesh_mask, 'lat':mesh_mask, 'depth':Wfiles[0],'time':Tfiles, 'data':Tfiles},
             'S': {'lon':mesh_mask, 'lat':mesh_mask, 'depth':Wfiles[0],'time':Sfiles, 'data':Sfiles}}

variables = {'U': 'vozocrtx',
             'V': 'vomecrty',
             'W': 'vovecrtz',
             'T': 'votemper',
             'S': 'vosaline'}

dimensions= {'U': {'lon':'glamf','lat':'gphif','depth':'depthw','time':'time_counter'},
             'V': {'lon':'glamf','lat':'gphif','depth':'depthw','time':'time_counter'}, 
             'W': {'lon':'glamf','lat':'gphif','depth':'depthw','time':'time_counter'},
             'T': {'lon':'glamt','lat':'gphit','depth':'depthw','time':'time_counter'},
             'S': {'lon':'glamt','lat':'gphit','depth':'depthw','time':'time_counter'}}


fieldset = FieldSet.from_nemo(filenames, variables, dimensions,allow_time_extrapolation=False,time_periodic=delta(days=6939))

```
thanks again!

**Reply by erikvansebille** [reply]
Hmm, very strange. I have no idea how even to start debugging this...

**Reply by ignasivalles** [reply]
Well, it's not a big deal the workaround (and maybe it's just a question of a conflict between packages in my conda workflow?). In any case it can take some time to find out what's going on, that's why I posted here, just in case someone experiences the same problem :)

**Reply by ignasivalles** [reply]
Hi again... after setting the time length of my dataset in `time_periodic`, I encountered the following issue while running parcels for a while (38% progress):

```
T sampled outside the time domain at time 1980-12-31T12:00:00.000000000. In fset.computeTimeChunk, try setting allow_time_extrapolation to True.
```

Here, T represents the temperature field that I'm sampling at each particle position, which is defined as U, V, and W in `fieldset.from_nemo`, as shown above.

I'm currently trying to figure out what is happening but any thoughts on this would be appreciated.

**Reply by ignasivalles** [reply]
My first time step in the dataset is 2001/01/01 and the last 2018/12/30 (daily frequency).

**Comment by ignasivalles** [comment]
Hi, following this discussion, I'm still confused about how the `time_periodic` and sample field work together (if they do). If I'm not wrong I've seen in previous posts that adding a field to be sampled does not implement the `time_periodic` functionality but maybe this was for previous versions. I also experience problems depending on how my input files are stored. When there is a file for each time step, there is no error. However, if my input files are a single file for each variable with full time dimension, I get the error previously mentioned `T sampled outside of the time domain`. 

In any case, what I'm doing now is setting `allow_time_extrapolation=True` for T and S before I add them to the `FieldSet` through the `add_field` function. So my question is: when `time_periodic` is on in the `FieldSet` for `UVW`, does using `allow_time_extrapolation=True`, for T and S without `time_periodic`, theoretically allow you to sample the fields correctly? even when the time index goes beyond the provided time range?

Maybe this is related with #1563 ?
